File: 1498-number-of-subsequences-that-satisfy-the-given-sum-condition/1498-number-of-subsequences-that-satisfy-the-given-sum-condition.java
class Solution {
    long mod = 1000000007;
    long powCal(long a, int n){
        long res = 1;
        while(n > 0){
            if(n%2==1) {
                res = (res * a)%mod;
                n--;
            }
            else {
                a = (a * 1L *a)%mod;
                n/=2;
            }
        }
        return res;
    }
    public int numSubseq(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        Arrays.sort(nums);
        int count = 0;
        while(l <= r){
            if(nums[l] + nums[r] > target) r--;
            else {
                count = (count + (int)powCal(2, r-l))%(int)mod;
                l++;
            }
        }
        return count;
    }
}

File: 1498-number-of-subsequences-that-satisfy-the-given-sum-condition/1498-number-of-subsequences-that-satisfy-the-given-sum-condition.cpp
class Solution {
    long long mod = 1e9+7;
    long long powCal(long long a, long long n){
        long long res = 1;
        while(n>0){
            if(n%2 == 1) {res = (res * a)%mod ; n--;}
            else {a = (a*a)%mod; n = n/2;}
        }
        return res%mod;
    }
    public:
    int numSubseq(vector<int>& nums, int target) {
        int n = nums.size();
        
        sort(begin(nums), end(nums));
        
        long long count = 0;
        int l = 0, r = n-1;
        while(l <= r){
            if(nums[l] + nums[r] > target) r--;
            else {
                count = (count + powCal(2, r - l))%mod; 
                l++;
            }
        }
        
        return count%mod; 
    }
};